### PR TITLE
feat(mfe): redact raw datasource queries on both /api/ds/query (frontend) and GET /dashboards/uid/{uid} (backend) when running as CR MFE

### PR DESCRIPTION
**Pull request by GurinderRawala**
## Summary

Two coordinated changes that remove raw datasource queries (SQL, PromQL, …) from the wire when Grafana is embedded as the CodeRabbit microfrontend (FNDashboard / CR MFE).

| # | Layer | What it does | Toggle |
|---|---|---|---|
| 1 | **Frontend** (`packages/grafana-runtime`) | Rewrites the `/api/ds/query` POST body to `{ dashboardUID, panelId, variables, filters, from, to }` instead of `{ queries, from, to }`. The backend resolves the saved query server-side. | `window.__FNDashboard__` (set by the MFE store) |
| 2 | **Backend** (`pkg/api`) | On `GET /api/dashboards/uid/{uid}`, walks the dashboard JSON and replaces known raw-query fields on every panel target with `"[REDACTED]"`. Recurses into row panels. | `GF_CR_MFE` env var (`1` / `true` / `yes` / `on`) |

Together these prevent raw datasource queries from ever reaching the browser when Grafana is embedded inside CodeRabbit, without changing standalone Grafana behaviour.

---

## Commit 1 — frontend: strip raw SQL from `/api/ds/query` body

`DataSourceWithBackend.query` in `packages/grafana-runtime` previously built a body of the form

```jsonc
{ "queries": [ { "rawSql": "…", "datasource": …, … } ], "from": "…", "to": "…" }
```

When `window.__FNDashboard__ === true` it now builds

```jsonc
{
  "dashboardUID": "<uid>",
  "panelId": <number>,
  "variables": { "<name>": "<value>", … },     // extracted from request.scopedVars
  "filters":   [ { key, operator, value }, … ], // ad-hoc filters
  "from": "<ms>",
  "to":   "<ms>"
}
```

### Why a window flag

`packages/grafana-runtime` cannot import from `public/app/*` (would create a cycle). Instead the existing MFE reducer (`public/app/store/configureMfeStore.ts`) mirrors its `state.FNDashboard = true` onto `window.__FNDashboard__`. The typing for the new window key lives in `public/app/types/window.d.ts`, which `packages/grafana-runtime/tsconfig.json` already pulls in.

### Files

- `packages/grafana-runtime/src/utils/DataSourceWithBackend.ts`
- `packages/grafana-runtime/src/utils/DataSourceWithBackend.test.ts` *(2 new unit tests)*
- `public/app/store/configureMfeStore.ts`
- `public/app/types/window.d.ts`

---

## Commit 2 — backend: redact queries from dashboard JSON when `GF_CR_MFE` is set

`GET /api/dashboards/uid/{uid}` (`hs.GetDashboard`) previously returned the dashboard JSON verbatim — every panel target included its raw `rawSql` / `expr` / `query` / `rawQuery` / `queryText` / `target` (graphite) string. When `GF_CR_MFE` is truthy, those string fields are now replaced with `"[REDACTED]"`. Everything else (panel layout, target metadata such as `refId` / `datasource` / `hide` / `format`, dashboard variables, …) is preserved so the frontend can still render the layout.

### Implementation notes

- The check follows the existing `GF_CR_*` env-var convention used elsewhere in the codebase (e.g. `GF_CR_POSTGRES_URL`, `GF_CR_RENDERER_AUTH_TOKEN_JWT_SECRET`).
- Recurses into row panels (which carry their children under a nested `panels` array).
- Only **string-valued** fields are masked, so e.g. influxdb's `rawQuery` boolean toggle keeps its type.
- The redaction lives in a separate file (`dashboard_mfe_mask.go`) so it's easy to extend with more field names if more datasource types are added later.

### Files

- `pkg/api/dashboard.go` — calls the helper in `GetDashboard` right after `dash.Data.Set("version", …)`.
- `pkg/api/dashboard_mfe_mask.go` — new helper + env-var detection.
- `pkg/api/dashboard_mfe_mask_test.go` — unit tests.

---

## Verification

| Command | Scope | Result |
|---|---|---|
| `yarn jest packages/grafana-runtime/src/utils/DataSourceWithBackend.test.ts --no-coverage` | frontend body rewrite | ✅ 15/15 pass (2 new) |
| `cd packages/grafana-runtime && yarn typecheck` | frontend types | ✅ no errors |
| `npx eslint <modified TS files>` | frontend lint | ✅ no errors |
| `go build ./pkg/api/...` | backend compile | ✅ |
| `go vet ./pkg/api/...` | backend static check | ✅ |
| `go test ./pkg/api/ -run "TestIsCodeRabbitMFE\|TestMaskDashboardQueriesForMFE" -count=1 -v` | new backend tests | ✅ 17/17 pass |
| `go test ./pkg/api/ -run "TestHTTPServer_GetDashboard_AccessControl\|TestDashboardAPIEndpoint" -count=1` | existing GetDashboard handler tests | ✅ |

### Pre-existing issues (NOT caused by this PR)

- Repository-wide `yarn typecheck` reports many errors inside `public/app/plugins/panel/volkovlabs-table-panel/**` — present on the base branch already, none of the modified files in this PR contribute new errors.
- `go test ./pkg/api/ -run TestDashboardSnapshotAPIEndpoint_singleSnapshot` panics with `no guardian factory implementation provided` — verified to also fail on the base branch with our changes stashed, so it is a pre-existing unrelated issue.

---

## Roll-out / env config

Self-hosted operators running Grafana as the CodeRabbit MFE should set:

```
GF_CR_MFE=1
```

on the Grafana backend process. The frontend half activates automatically when the MFE store enters dashboard mode.

## Assumptions

- `variables` (frontend body) is built from `request.scopedVars` by extracting each `ScopedVar.value`.
- `filters` (frontend body) is forwarded verbatim from `request.filters`, defaulting to `[]`.
- The masked frontend field list — `rawSql`, `expr`, `query`, `rawQuery`, `queryText`, `target` — covers the built-in Grafana datasources. If new datasource types are introduced, add their raw-query field to `crMaskedQueryFields` in `pkg/api/dashboard_mfe_mask.go`.
- The window flag is not unset on tear-down — once a host enters MFE mode it stays there for the SPA lifetime, matching the existing `state.FNDashboard` behaviour.

## Summary by CodeRabbit

* **New Features**
  * Added FNDashboard (microfrontend) support to enrich datasource query requests with FN context (templating variables + dashboard UID) without exposing raw query text.
  * Extended microfrontend masking to redact dashboard query fields and to scrub executed-query strings in `/api/ds/query` responses.
* **Bug Fixes**
  * Improved dashboard UID fallback across request, rendering UID, and URL path.
  * Ensured variable updates trigger full refresh appropriately in FNDashboard mode.
* **Tests**
  * Expanded unit and end-to-end coverage for FN request shaping, redaction/masking, and edge cases (including unset/nil/empty).

### CODE REVIEW

**Comment by GurinderRawala** [comment]
### Follow-up commit: forward dashboard-level template variables in the FN body

Debugging a local environment showed every panel returning "no data" once the proxy resolver was wired up. The proxy logs revealed the FN body's `variables` payload only contained `__interval` / `__interval_ms`:

```
[Grafana proxy]:: FN body received {
  dashboardUID: 'timeMetrics', panelId: 77,
  variableKeys: [ '__interval', '__interval_ms' ],
  ...
}
```

Cause: in `DataSourceWithBackend.query` I built `variables` from `request.scopedVars`, but `scopedVars` only carries panel-local vars (`__interval`, repeat). Dashboard-level template variables (`org_id`, `repo_name`, `self_hosted_id`, ...) live on `TemplateSrv`. Without them the proxy substitutes every `$org_id` / `${repo_name:sqlstring}` with the dashboard's `allValue` (`''`) and the SQL legitimately returns zero rows.

Fix:

- Read `getTemplateSrv().getVariables()` and seed the FN body's `variables` map from each `current.value`.
- Panel-scoped vars from `request.scopedVars` still override dashboard-level vars of the same name (so repeat variables and `__interval` keep working).
- Lookup wrapped in `try/catch` since `TemplateSrv` is undefined in tests / non-dashboard contexts.

Test `merges dashboard-level template variables into the FN body` added; existing FN body test still passes (5 snapshots, 16 tests total green).

Operators running the local Grafana fork need to rebuild the frontend bundle for the change to take effect — the dev `tsx watch` only restarts the Go backend.

**Comment by GurinderRawala** [comment]
### Follow-up commit: mask templating-variable SQL as well

The first masking commit only redacted SQL on `panels[*].targets[*]`, but every dashboard's templating variable list (`templating.list[*]` of type `query`) also carries the raw SQL in its `query` and `definition` fields — used by the variable picker to populate dropdown options. So simply opening a dashboard as the CR microfrontend still leaked the SQL of every filter / variable definition.

Probed a real shipped dashboard (`summary.json`) before/after the new mask:

```
name=self_hosted_id    type=custom    query=""           definition=""             // untouched
name=org_id            type=custom    query=""           definition=""             // untouched
name=org_name          type=query     query="[REDACTED]" definition="[REDACTED]"   // masked
name=repo_name         type=query     query="[REDACTED]" definition="[REDACTED]"   // masked
name=username          type=query     query="[REDACTED]" definition="[REDACTED]"   // masked
name=teams             type=query     query="[REDACTED]" definition="[REDACTED]"   // masked
```

### What changed

- `maskDashboardQueriesForMFE` now walks `templating.list[*]` in addition to `panels`.
- For each variable of type `query` it redacts `definition` (always a string) and `query` — either as a string (legacy / our shipped dashboards) or, if it's an object (Grafana 9+ UI-edited form), by reusing the same raw-query field masker we already apply to panel targets.
- `custom`, `textbox`, `interval`, `datasource` variables are skipped — they don't carry datasource SQL and the picker depends on `current` / `options` instead.
- Empty strings are left intact (so the shape stays recognisably "no query" rather than a misleading `[REDACTED]`).
- Factored the per-object field redaction into a small `maskRawQueryFields` helper.

### Verification

| Command | Result |
|---|---|
| `go build ./pkg/api/...` | ✅ |
| `go vet ./pkg/api/...` | ✅ |
| `go test ./pkg/api/ -run "TestIsCodeRabbitMFE\|TestMaskDashboardQueriesForMFE"` | ✅ **19/19 pass** (2 new) |
| `go test ./pkg/api/ -run "TestHTTPServer_GetDashboard_AccessControl\|TestDashboardAPIEndpoint"` | ✅ (existing handler tests still pass) |

Commit: `194b2c1075b`.

**Comment by GurinderRawala** [comment]
### Follow-up commit: emit a `variableQueries` FN body when `rawSql` is masked

The backend's templating-var masking flips every `query`-type variable's SQL to `"[REDACTED]"`. Grafana's variable runner then dispatches that SQL through `DataSourceWithBackend.query` to `/api/ds/query` — and the proxy was rejecting it with **400** because `[REDACTED]` doesn't match any shipped template.

`DataSourceWithBackend.query` now detects this case (any target with `rawSql === "[REDACTED]"`) and emits a new FN body shape instead:

```jsonc
{
  "dashboardUID": "<uid>",
  "variableQueries": [{ "refId": "<variable name>" }, ...],
  "variables":     { ... },
  "filters":       [ ... ],
  "from":          "<ms>",
  "to":            "<ms>"
}
```

`request.dashboardUID` is unset for variable queries, so the MFE store reducer now also mirrors the current rendering dashboard UID to `window.__FNDashboardRenderingUID__` (declared in `window.d.ts`); `DataSourceWithBackend` prefers `request.dashboardUID` and falls back to the window value.

The companion proxy change (coderabbitai/mono#20127) resolves `refId → SQL` from the indexed dashboard JSON, runs the same variable interpolation + macro expansion + datasource resolution it already does for panel queries, and forwards a legacy `{ queries: [...] }` body to upstream Grafana.

#### Verification

| Command | Result |
|---|---|
| `yarn jest packages/grafana-runtime/src/utils/DataSourceWithBackend.test.ts` | ✅ **17/17 pass** (1 new: `emits a variableQueries body when rawSql is [REDACTED] and panelId is missing`) |
| `cd packages/grafana-runtime && yarn typecheck` | ✅ no errors |
| `npx eslint <modified files>` | ✅ 0 errors |

Commit: `6f3deb03032`.

**Comment by GurinderRawala** [comment]
### Follow-up commit: discriminate FN body on `panelId`, not on redacted `rawSql`

The previous heuristic ("any target with `rawSql === '[REDACTED]'` ⇒ variableQueries body") was wrong: after the backend mask, **panel** targets *also* arrive with `rawSql === '[REDACTED]'`. So every panel query got routed into the variable branch; the proxy then tried to resolve `refId="A"` against the templating list, failed, and **400'd every panel**.

Switched the discriminator to `request.panelId`:

| Runner | Sets `panelId`? | Body shape |
|---|---|---|
| `PanelQueryRunner` | ✅ numeric | `FnPanelQueryRequestBody` |
| `VariableQueryRunner` | ❌ undefined | `FnVariableQueryRequestBody` |

The redaction itself is incidental — Grafana already gives us a clean discriminator, we just weren't using it.

Regression test added: `panel queries with redacted rawSql still emit a panel body (not variableQueries)`.

| Command | Result |
|---|---|
| `yarn jest packages/grafana-runtime/src/utils/DataSourceWithBackend.test.ts` | ✅ **18/18 pass** (1 new) |
| `cd packages/grafana-runtime && yarn typecheck` | ✅ |
| `npx eslint <modified files>` | ✅ 0 errors |

Commit: `1e6341ab7da`.

**Comment by GurinderRawala** [comment]
### Refactor: encode panel/variable identifier in the redacted SQL itself

Replaced the opaque `"[REDACTED]"` placeholder with a structural key:

```
panel target:        [CR_REDACTED:p:<panelId>:<refId>]
templating variable: [CR_REDACTED:v:<variableName>]
```

The key carries everything the proxy needs to find the original SQL. This collapses three earlier complications down to nothing:

- ❌ No more panel-vs-variable FN body discriminator (`isVariableQuery` heuristic).
- ❌ No more `window.__FNDashboardRenderingUID__` mirror for the body shape (we still read it as a *fallback* for the sidecar's `dashboardUID` when `request.dashboardUID` is unset — e.g. for variable queries dispatched by `VariableQueryRunner`).
- ❌ No more parallel `variableQueries` request shape.

The frontend now just forwards the legacy `{ queries, from, to }` body verbatim, attaching a small `crFnContext: { variables, filters, dashboardUID }` sidecar. The proxy parses each query's `rawSql`, decodes the key, looks the SQL up, substitutes variables / macros, and rewrites the query in place. Grafana's upstream datasource silently ignores unknown body fields, so the same body is forwarded after resolution.

Companion proxy change: coderabbitai/mono#20127 commit `bc2c178693`.

#### Verification

| Command | Result |
|---|---|
| `go test ./pkg/api/ -run "TestMaskDashboardQueriesForMFE\|TestIsCodeRabbitMFE"` | ✅ all sub-tests pass (assertions updated for new keys) |
| `yarn jest packages/grafana-runtime/src/utils/DataSourceWithBackend.test.ts` | ✅ 16/16 pass (FN dispatch tests rewritten to assert `crFnContext`) |
| `cd packages/grafana-runtime && yarn typecheck` | ✅ |
| `npx eslint <modified files>` | ✅ 0 errors |
| Live probe on `summary.json` after restart with `GF_CR_MFE=1` | ✅ panel rawSql → `[CR_REDACTED:p:102:A]`, templating queries → `[CR_REDACTED:v:org_name]` etc. |

Commit: `243abfcb13b`.

**Comment by GurinderRawala** [comment]
### Follow-up: attach `crFnContext` from `metricFindQuery` too + robust `dashboardUID` fallback

Live testing on the user's local environment revealed two remaining gaps after the structural-key redaction redesign:

1. **Templating-variable queries** (the dropdown population queries fired by `VariableQueryRunner` at dashboard-init time) go through `SqlDatasource.metricFindQuery → runMetaQuery`, which posts directly to `/api/ds/query` via `getBackendSrv().fetch` and **completely bypasses `DataSourceWithBackend.query`**. They were arriving at the proxy with no `crFnContext` sidecar at all.
2. **Inside Qiankun's sandbox** the MFE runs in an `about:blank` iframe whose own `location.pathname === "blank"`, so the URL-based `dashboardUID` fallback couldn't see the real `/dashboard/<uid>` path.

#### What changed

- New shared helper `packages/grafana-runtime/src/utils/fnDashboardBody.ts` exporting `buildCrFnContext` (variables + filters + dashboardUID), `hasRedactedRawSql`, and `isFnDashboardWindow`. Re-exported from the `@grafana/runtime` barrel.
- `SqlDatasource.runMetaQuery` (`packages/grafana-sql`) now attaches `crFnContext` when any of its queries carries a redacted `rawSql`. metricFindQuery / variable queries finally get the same treatment as panel queries.
- `DataSourceWithBackend.query` refactored to use the same helper. The dispatch check is now **"any redacted target OR FN window flag"** rather than the FN window flag alone — variable queries dispatched at init time arrive *before* the MFE store sets `__FNDashboard__`, but they always carry a redacted `rawSql`.
- `buildCrFnContext`'s `dashboardUID` resolution walks the frame chain: `request.dashboardUID → window.__FNDashboardRenderingUID__ → window.location.pathname → window.parent.location.pathname → window.top.location.pathname`. The last two hops bypass Qiankun's `about:blank` sandbox iframe.

#### End-to-end verification on the running UI

After rebuilding the frontend bundles (including the postgres plugin's `dist/module.js`) and reloading `http://localhost:5173/dashboard/summary?var-org_id=fef3aa20-...`, the dashboard renders **real data**:

| Stat | Value |
|---|---|
| Active Repositories | 13 |
| Merged Pull Requests | 118 (avg 1.93 / user) |
| Active Users | 43 / 16 |
| Chat Usage | 11.0k / 31.8k |
| Median Time | 2.87 min |
| Reviewer Time Saved | 1.02 days |
| CodeRabbit Review Comments | 25 |
| Severity Distribution | Critical 40%, Major 52%, Minor 4%, Trivial 4% |

Variable dropdowns populate; proxy log has **no FN-redacted errors** anymore — every `/api/ds/query` proxy line is a successful resolution.

#### Test verification

| Command | Result |
|---|---|
| `yarn jest packages/grafana-runtime/src/utils/DataSourceWithBackend.test.ts` | ✅ 17/17 pass |
| `cd packages/grafana-runtime && yarn typecheck` | ✅ |
| `cd packages/grafana-sql && yarn typecheck` | ✅ |

Commit: `f69af27ae7f`.

**Comment by GurinderRawala** [comment]
### Follow-up: refresh all panels on variable change when masking is active

Live testing revealed a third FN-mode bug after the structural-key redaction landed: **filter dropdowns updated the URL but did not refresh the panels.** Selecting `howonlee` in the Username picker would update `?var-username=howonlee` and the picker chip, but every panel still showed the unfiltered numbers.

#### Root cause (instrumented in-browser)

Grafana decides which panels to re-fetch on a variable change via:
1. `getPanelVars(panels)` — walks each panel JSON looking for literal `${var}` references and builds `panelsByVar: Record<varName, Set<panelId>>`.
2. `getAllAffectedPanelIdsForVariableChange([variableId], graph, panelsByVar)` — returns `Set<panelId>`.
3. `variableUpdated` publishes `new VariablesChanged({ refreshAll: false, panelIds, … })`.
4. `DashboardModel.variablesChangedHandler` calls `startRefresh(payload)` which re-queries only the listed panels.

With the backend mask in place, every panel target's `rawSql` is `[CR_REDACTED:p:<panelId>:<refId>]` — **no `${var}` references remain in the JSON**. `getPanelVars` therefore returns an empty map for every variable, `getAllAffectedPanelIdsForVariableChange` returns `Set()`, and `startRefresh({ panelIds: [], refreshAll: false })` is a no-op.

Console proof (with temporary debug logs):

```
[FN DEBUG] panels for variable change {panelCount: 15, panelVarsKeys: 7, varName: username, varId: username}
[FN DEBUG] publishing VariablesChanged {refreshAll: false, panelIds: Array(0), variable: Object}
[FN DEBUG] DashboardModel.variablesChangedHandler {refreshAll: false, panelIds: Array(0), variable: Object}
```

15 panels exist, 7 variable-name buckets, but the affected-panel set is empty.

#### Fix

In `variableUpdated`, when `window.__FNDashboard__` is set, fall back to `refreshAll: true` — same treatment as ad-hoc variables, which already use this path because we never know which panels they affect.

```diff
+    const isFnDashboard =
+      typeof window !== 'undefined' && (window as any).__FNDashboard__ === true;
     const event: VariablesChangedEvent =
-      variableInState.type === 'adhoc'
+      variableInState.type === 'adhoc' || isFnDashboard
         ? { refreshAll: true, panelIds: [] }
         : {
             refreshAll: false,
             panelIds: Array.from(getAllAffectedPanelIdsForVariableChange([variableInState.id], g, panelVars)),
             variable: getVariable(identifier, state),
           };
```

#### Live verification

After rebuilding the bundle and reloading:

| Action | Stat | Before fix | After fix |
|---|---|---|---|
| `?var-username=All` | Active Repos | 7 | 7 |
| `?var-username=All` | Chat Sessions | 2.76k | 2.76k |
| Pick `howonlee` in picker | Active Repos | 7 (stale) | **3** (refreshed) |
| Pick `howonlee` in picker | Chat Sessions | 2.76k (stale) | **2** (refreshed) |
| Reset to All in picker | Active Repos | 3 (stale) | **7** (restored) |

Variable dropdowns now drive panel refreshes correctly end-to-end.

Commit: `be35bf0ece4`.

File: pkg/api/dashboard_mfe_mask_test.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+	}
+
+	t.Run("unset", func(t *testing.T) {
+		// t.Setenv only restores; we want to test the missing-key path explicitly.
+		// The "" case above already sets it to empty (still present), so verify
+		// the LookupEnv-not-ok path by relying on the default test environment
+		// where the var is not set.
+		// (No-op sanity assertion — the value above covers the truthy parsing.)
+		assert.False(t, isCodeRabbitMFE() && false)
+	})
```
[the remark this answers is not in the document]
Comment (GurinderRawala) [reply pkg/api/dashboard_mfe_mask_test.go:56]: Valid — fixed in `77ad9c8b7e9`. The `unset` subtest now snapshots the env var, calls `os.Unsetenv(crMFEEnvVar)`, asserts `isCodeRabbitMFE()` returns false, and restores the original value via `t.Cleanup`. Added `"os"` to the imports.

File: pkg/api/dashboard_mfe_mask.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+	crMaskPrefix = "[CR_REDACTED:"
+	crMaskSuffix = "]"
+)
+
+func panelMaskValue(panelID int64, refID string) string {
+	return crMaskPrefix + "p:" + strconv.FormatInt(panelID, 10) + ":" + refID + crMaskSuffix
+}
+
+func variableMaskValue(name string) string {
+	return crMaskPrefix + "v:" + name + crMaskSuffix
```
[the remark this answers is not in the document]
Comment (GurinderRawala) [reply pkg/api/dashboard_mfe_mask.go]: Valid — fixed in `77ad9c8b7e9` (Grafana) and `d4c02593fd` (mono proxy).

- `panelMaskValue` / `variableMaskValue` now wrap `refID` / `name` in `url.QueryEscape` before concatenation.
- The proxy's `parseMaskKey` decodes each segment via `decodeURIComponent` (via a small `safeDecodeURIComponent` helper that catches malformed escapes and returns `null`).
- For names containing only `[A-Za-z0-9_-]` — every refId / variable name in the shipped dashboards — the encode/decode is a no-op, so existing assertions like `[CR_REDACTED:p:1:A]` / `[CR_REDACTED:v:org_name]` still match exactly.
- New regression tests:
  - **Grafana**: `URL-encodes delimiter characters in refId and variable name` — confirms `A:B]C` → `A%3AB%5DC` and `weird:name]thing` → `weird%3Aname%5Dthing`.
  - **Proxy**: `resolveRedactedQueries decodes percent-escaped refId / variable names` + `malformed percent escape in mask payload fails closed`. The second test also tightens `resolveSingleQuery` so a rawSql that starts with `[CR_REDACTED:` but fails `parseMaskKey` returns `null` (fail closed) instead of forwarding the literal placeholder to upstream Grafana.

File: packages/grafana-sql/src/datasource/SqlDatasource.ts [hunk 3]
```diff
@@ -231,17 +234,27 @@ export abstract class SqlDatasource extends DataSourceWithBackend<SQLQuery, SQLO
...
+    // rawSql is a redaction key (e.g. `[CR_REDACTED:v:org_name]`).
+    const rawSqls = queries.map(q => (q as unknown as { rawSql?: unknown }).rawSql as string | undefined);
+    const data: Record<string, unknown> = {
+      from: range.from.valueOf().toString(),
+      to: range.to.valueOf().toString(),
+      queries,
+    };
+    if (isFnDashboardWindow() || hasRedactedRawSql(rawSqls)) {
+      data.crFnContext = buildCrFnContext({});
+    }
```
[the remark this answers is not in the document]
Comment (GurinderRawala) [reply packages/grafana-sql/src/datasource/SqlDatasource.ts]: Valid — fixed in `77ad9c8b7e9`.

- `metricFindQuery` now passes the `scopedVars` it already assembles (`options.scopedVars` + `getSearchFilterScopedVar`) into `runMetaQuery`.
- `runMetaQuery` accepts an optional `scopedVars: ScopedVars` and forwards it via `buildCrFnContext({ scopedVars })`.

This means redacted variable SQL keeps access to `__searchFilter` and any other panel-scoped overrides during proxy interpolation — the proxy already merges these on top of the dashboard-level template vars in `buildCrFnContext`.

File: public/app/features/variables/state/actions.ts [hunk 4]
```diff
@@ -643,8 +643,18 @@ export const variableUpdated = (
...
+    // When Grafana is running as the CodeRabbit microfrontend the backend
+    // masks every panel's rawSql to a structural key like
+    // `[CR_REDACTED:p:<panelId>:<refId>]`. The dependency graph in
+    // `getPanelVars` keys off `${var}` references inside each panel's JSON
+    // and so sees zero affected panels for any variable change, breaking
+    // the dashboard refresh. Fall back to `refreshAll: true` in this mode
+    // so every panel re-renders on every variable change — same treatment
+    // as ad-hoc variables.
+    const isFnDashboard =
+      typeof window !== 'undefined' && (window as any).__FNDashboard__ === true;
```
[the remark this answers is not in the document]
Comment (GurinderRawala) [reply public/app/features/variables/state/actions.ts]: Valid — fixed in `77ad9c8b7e9`. Dropped the `(window as any)` cast; `window.__FNDashboard__` is already declared on the `Window` interface in `public/app/types/window.d.ts` from PR-1, so the typed access works without any escape hatch.

**Review by abhilash-srivathsa — APPROVED** [review]
